mtt/src/mtt_types.rs

```diff
@@ -118,3 +116,226 @@ impl MttL1EntryType {
+    let mtt_index = (addr & L2_ADDR_MASK) >> L2_ADDR_SHIFT;
+    let mtt_entry_addr = (get_mtt_base() + mtt_index) as *const u64;
+    assert!(mtt_index <= (8 * 1024 * 1024) - 8);
+    // Safety: The MTT is a 8MB contiguous region allocated by trusted platform-FW. The TSM is
+    // the only entity responsible for maintaining it, and we are using a 23-bit index into it
+    let value = unsafe { *mtt_entry_addr };
+    if let Some(l2_entry) = MttL2EntryType::from_raw(value) {
+        l2_entry
+    } else {
+        panic!("MTT L2 entry for {addr:x} was invalid (value:x)")
+    }
+}
+
+fn set_l2_entry(phys_addr: SupervisorPageAddr, entry: MttL2EntryType) {
+    let addr = phys_addr.bits();
+    let mtt_index = (addr & L2_ADDR_MASK) >> L2_ADDR_SHIFT;
+    assert!(mtt_index <= (8 * 1024 * 1024) - 8);
+    let mtt_entry_addr = (get_mtt_base() + mtt_index) as *mut u64;
+    // Safety: The MTT is a 8MB contiguous region allocated by trusted platform-FW. The TSM is
+    // the only entity responsible for maintaining it, and we are using a 23-bit index into it
+    unsafe { *mtt_entry_addr = entry.raw() };
+}
+
+// The L1 is a 4KB page that's allocated by the TSM to manage the memory type (non-confidential or confidential) at a 4K-page granularity.
+// This is done by dividing the 64MB region containing the page into 16KB regions of 4K each, and the 2-bit encoding in the L1 page pointed
+// to by the L2 entry determines the type.
+// The 12-bit index into the 4K page is computed using the 14-bits [25:12] of the physical address, and then shifting right by 2
+#[allow(dead_code)]
+fn get_l1_entry(phys_addr: SupervisorPageAddr) -> MttL1EntryType {
+    let addr_bits = phys_addr.bits();
+    let l2_entry = get_l2_entry(phys_addr);
+    match l2_entry {
+        MttL1Dir(pfn) => {
+            let (l1_index, sub_index) = get_l1_index_and_subindex(phys_addr);
+            let value = get_l1_value(pfn, l1_index);
+            if let Some(l1_entry) = MttL1EntryType::from_raw((value >> sub_index) & 0x3) {
+                l1_entry
+            } else {
+                panic!("L1 entry encoding for {addr_bits:x} was invalid {value:x}");
+            }
+        }
+        _ => panic!("Attempt to get MttL1Dir from {l2_entry:?} for {addr_bits:x}"),
+    }
+}
+
+fn get_l1_index_and_subindex(phys_addr: SupervisorPageAddr) -> (u64, u64) {
+    let addr_bits = phys_addr.bits();
+    // Bits 25:17 for the 9-bit index, then 3-bit left shift for the 8-byte entry
+    let l1_index = (addr_bits & 0x03fe_0000) >> 14;
+    // 5-bit sub-index (16:12) with 1-bit left shift for the 2-bits per entry
+    let sub_index = (addr_bits & 0x1_f000) >> 11;
+    (l1_index, sub_index)
+}
+
+fn get_l1_value(pfn: u64, l1_index: u64) -> u64 {
+    let mtt_l1_page = pfn << 12;
+    let l1_entry_addr = (mtt_l1_page + l1_index) as *const u64;
+    assert!(l1_index <= 4096 - 8);
+    // Safety: The 4K L1 page is allocated by the TSM, and we are using a 12-bit index.
```

Comment:
how are these guarantees enforced? the function accepts an arbitrary pfn and reads it. How does this code know the pfn isn't aliased elsewhere?